### ISSUE TITLE
Produce lint violations based on dry-run of format without changing the original source code

### ISSUE THREAD

**Issue by paul-dingemans** [pinterest/ktlint#2086]
Currently `lint` reports the violations without changing the AST. As a result, violations are reported which do not match with the code produced with `format`.

Given code below:
```
fun foo(a: String, b: String, c: String,) { return a + b + c }
```

Linting above code results (depending on your configuration) into following violations:
```
Foo.kt:1:9: Newline expected after opening parenthesis (standard:function-signature)
Foo.kt:1:20: Parameter should start on a newline (standard:function-signature)
Foo.kt:1:31: Parameter should start on a newline (standard:function-signature)
Foo.kt:1:40: Whitespace after ',' is missing (standard:parameter-list-spacing)
Foo.kt:1:40: Unnecessary trailing comma before ")" (standard:trailing-comma-on-declaration-site)
Foo.kt:1:41: Newline expected before closing parenthesis (standard:function-signature)
```

Note that the violation `Unnecessary trailing comma before ")" (standard:trailing-comma-on-declaration-site)` is actually incorrect or redundant when the other violations are actually fixed as can be seen in code below produced by `format`:
```
fun foo(
    a: String,
    b: String,
    c: String,
) { return a + b + c }
```

When `lint` changes the AST in the same way as `format` but without returning the changed source code, the violations will be aligned with the code that actually is produced by `format`. Also, this clear lots of unnecessary if-checks in the rules as it is no longer necessary to check the `autocorrect` variable.
```
// Given that some violation is found
emit(someOffset, "some details", true)
if (autoCorrect) {
    // actuall correct the error
}
```
becomes:
```
// Given that some violation is found
emit(someOffset, "some details", true)
// actuall correct the error
```

**Comment by paul-dingemans** [comment]
When implementing this change, it turned out that it have unintended side-effects in the user experience of `lint`. This can best be made clear with an example unit test of the original code:
```
@Test
fun `Given it is nested in a THEN`() {
    val code =
        """
        fun foobar() {
            if (true) if (false) foo() else bar()
        }
        """.trimIndent()
    val formattedCode =
        """
        fun foobar() {
            if (true)
                if (false)
                    foo()
                else
                    bar()
        }
        """.trimIndent()
    ifElseWrappingRuleAssertThat(code)
        .hasLintViolations(
            LintViolation(2, 15, "Expected a newline"),
            LintViolation(2, 26, "Expected a newline"),
            LintViolation(2, 32, "Expected a newline"),
            LintViolation(2, 37, "Expected a newline"),
        ).isFormattedAs(formattedCode)
}
```
Line 2 contains multiple errors according to the `if-else-wrapping` rule. Each of the offsets point to a relevant location in the original code.

If `lint` would alter the AST during `lint`, the offsets of the lint violation become dependent on changes made while fixing lint violations which were detected before. The assertions of the previous unit test would need to be altered as follows to make the test pass:
```
ifElseWrappingRuleAssertThat(code)
    .hasLintViolations(
        LintViolation(2, 15, "Expected a newline"),
        LintViolation(3, 20, "Expected a newline"),
        LintViolation(4, 19, "Expected a newline"),
        LintViolation(5, 18, "Expected a newline"),
    ).isFormattedAs(formattedCode)
```
Although the offets above are correct when all preceding violations are fixed as suggest, they do not make sense when comparing them with the orginal source code. Also, in the unit test it would become confusing to use the altered offsets.
